### ISSUE TITLE
One CPU core always showing 100%

### ISSUE THREAD

**Issue by ghost** [fflewddur/tophat#107]
I'm running TopHat 11 on Pop_OS! on an AMD Ryzen 7 6800H. Previously TopHat worked fine on my system and while this may be an issue on my end, I can't seem to figure it out.

TopHat is always reporting one CPU core out of 16 (8 hyperthreaded) always being at 100%, despite other tools like gnome-system-monitor and htop showing 1-2% usage.

I've set htop to not hide kernel threads and still I don't see hardly any usage... even TopHat itself doesn't actually show any application using much CPU.

![image](https://github.com/fflewddur/tophat/assets/60193408/6f670cd1-da81-4e9b-b776-d3d44489f9d9)
![image](https://github.com/fflewddur/tophat/assets/60193408/5b45cf41-246c-4a2b-9b28-f9d28e87a440)

**Comment by ryryog25** [comment]
Not sure if this is technically a me issue, but I think this is because however TopHat is getting info, it's the same as setting htop to include detailed usage, which shows that core with a gray full bar, and this seems linked to IRQs

**Comment by mattia-b89** [comment]
I have never had such kind of issue.
Is it still valid for you?

**Comment by webdevelopers-eu** [comment]
I have configured the extension to display only CPU and Network metrics, yet it keeps a one CPU busy at 11%. Maybe this is the same issue?

Upon investigation, it appears that the extension frequently polls all stats from /proc/*, which likely results in unnecessary processing and continuous load. Ideally, polling should occur only when triggered by a user action, such as clicking on the tray icon, at which point the detailed information is actually needed. Reducing polling frequency or implementing an on-demand approach could significantly lower the extension's resource usage.

**Comment by fflewddur** [comment]
@ghost, yes, TopHat includes IRQ usage in its process calculations, so what you're seeing is intentional. Do you know why one of your cores is constantly showing high IRQ usage?

@webdevelopers-eu, not quite. TopHat calculates usage over time (3 seconds by default, but this is user configurable) and TopHat itself certainly shouldn't be driving that much CPU usage. I would love to limit the /proc/* reads to only occur when the detail menus are displayed, but this creates a different problem: when I see a CPU or memory spike, I want to click on the menu to immediately see what caused it. But, /proc/* record counters are always incrementing for things like CPU usage and disk activity--this means we need to know the state of the CPU, say, 3 seconds ago, and compute how much it changed during that time period. If those details are only ready from /proc/* when the menu is open, then we need to wait for multiple polls of /proc/*, so the menu would be empty for a few seconds, and wouldn't reflect whatever usage drove the spike that caused me to look at it in the first place. I'm all ears if you have any ideas for how to more efficiently solve this problem.

**Comment by fflewddur** [comment]
I've just released TopHat v18, which includes additional performance improvements. Is anyone still having this issue with the latest release?

**Comment by rome-legacy** [comment]
Happy new year, fratellis,
i have just installed ghostty on my ubuntu 24.04.1 with TopHat V18 and noticed that two of my X Cores are shown as 100%, but my CPU temperature was not affected. When i exit ghostty, back to normal. That made wonder.
So i have enabled percentage and its low. This lead me to this issue.

![Screenshot from 2025-01-08 23-08-18](https://github.com/user-attachments/assets/ae420e38-0340-4d97-ab1f-da17ff46fb63)
So i would say the problem is still there.

If I open htop and enable display option "Detailed CPU Time (System/IO-Wait/Hard-IRQ/Soft-IRQ/Steal/Guest)" i can also see "higher" bars there.

kind regards
rome

**Comment by fflewddur** [comment]
@rome-legacy That doesn't sound like a bug. Whether those cores are busy executing user-space code, handling IRQs, waiting for I/O, etc., doesn't matter because they're still completely in use and unavailable for any other task. If a particular application consistently causes a spike in CPU use, it's likely either because that application is very resource-hungry, or it's not behaving as expected. That second case is exactly what TopHat is intended to help people identify.

**Comment by rome-legacy** [comment]
Ok, thx for the info 👍

**Comment by schopin-pro** [comment]
Hi,

FWIW, I'm on version 19 and am still seeing the activity spikes on htop, where gnome-shell spikes to 50% of a CPU every 3 seconds when the CPU monitor is active.

Personally, I'd be fine with coarse monitoring of global CPU usage and only start collecting process data when actively clicking on the monitor. Active monitoring could even be done with a higher refresh rate.
